[sources] Report indefinite errors from the postgres source

bkirwi opened the pull request:
### Motivation

Known-desired: https://github.com/MaterializeInc/materialize/issues/16065

### Tips for reviewer

I feel bad about the third commit on this branch, but I also think it's better than doing nothing? Feel free to tell me it's a bad idea, though.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

aljoscha commented on src/storage/src/source/postgres.rs:
```diff
@@ -428,12 +428,25 @@ async fn postgres_replication_loop_inner(
+                // above will make it to the status history table, so we'll wait a little
+                // bit before halting. This should be removed once we have a more graceful
+                // shutdown approach.
+                tokio::time::sleep(Duration::from_secs(10)).await;
```
aha! I see what you mean: the path that the above status update takes is quite long, it goes out to the thing that drives the `SourceReader`, from there goes across timely channels to a different operator, and only there is it written into the status shard. And all that has to happen before we `halt!` here because that stops the whole process.

Though one... :see_no_evil:

Strawman: write to the status shard right here, before doing the `halt!`?

bkirwi replied:
Can't do it I think! We don't have access to persist here, and it looks like it would be a challenge to thread all the way through.

If we're nervous about the sleep -- and I think that's a reasonable stance -- I should probably just revert the commit and eat the slightly worse error reporting.

aljoscha reviewed (commented):
Everything but the halting problem (ha!) seems fine, I had one strawman idea but that might also not be very good.

bkirwi commented:
Rebased and reverted the sketchy commit, just so it doesn't hold up the less controversial stuff - I can follow up with whatever approach we settle on later this week.
